File: packages/python/convai_analytics/resources/latency.py
"""Latency facade — convenience wrappers over breakdown/timeseries."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..measures import GroupBy, MetricNames, Segments, raw_value_percentile_measure
from ..types import BreakdownResponse, TimeseriesResponse

if TYPE_CHECKING:
    from ..client import ConvaiAnalytics


class LatencyFacade:
    def __init__(self, client: ConvaiAnalytics) -> None:
        self._client = client

    def by_component(
        self,
        *,
        percentile: str = "p95",
        **filters: Any,
    ) -> BreakdownResponse:
        """"Which component contributes most to my p95 end-to-end latency?"

        Delegates to ``breakdown(measure='p95Value', group_by='processor',
        segment='endToEndTurnLatency')``.
        """
        return self._client.breakdown(
            measure=raw_value_percentile_measure(percentile),  # type: ignore[arg-type]
            group_by=GroupBy.PROCESSOR,
            segment=Segments.END_TO_END_TURN_LATENCY,
            **filters,
        )

    def over_time(
        self,
        *,
        percentile: str = "p95",
        granularity: str | None = None,
        **filters: Any,
    ) -> TimeseriesResponse:
        """"Plot p95 end-to-end latency over time."

        Delegates to ``timeseries(measure='p95Value',
        metric_name='voice.user_to_bot_latency')``.
        """
        return self._client.timeseries(
            measure=raw_value_percentile_measure(percentile),  # type: ignore[arg-type]
            metric_name=MetricNames.VOICE_USER_TO_BOT_LATENCY,
            granularity=granularity,
            **filters,
        )

File: packages/typescript/src/resources/latency.ts
/**
 * LatencyFacade — convenience wrappers over `breakdown` and `timeseries`
 * for the most common latency questions. Pure delegation; the named REST
 * surface is the source of truth.
 */

import type { ConvaiAnalytics } from "../client.js";
import {
  GROUP_BY,
  METRIC_NAMES,
  rawValuePercentileMeasure,
  SEGMENTS,
} from "../measures.js";
import type {
  BreakdownResponse,
  CommonFilters,
  FirstResponseBreakdownParams,
  FirstResponseBreakdownResponse,
  FirstResponseSummaryParams,
  FirstResponseSummaryResponse,
  FirstResponseTimeseriesParams,
  FirstResponseTimeseriesResponse,
  Percentile,
  TimeRange,
  TimeseriesResponse,
} from "../types.js";

export interface LatencyByComponentParams extends TimeRange, CommonFilters {
  /** Default p95. */
  percentile?: Percentile;
}

export interface LatencyOverTimeParams extends TimeRange, CommonFilters {
  /** Defaults to ['p50', 'p95', 'p99']. */
  percentiles?: Percentile[];
  granularity?: "minute" | "hour" | "day";
}

export class LatencyFacade {
  readonly #client: ConvaiAnalytics;

  constructor(client: ConvaiAnalytics) {
    this.#client = client;
  }

  /**
   * "Which component contributes most to my p95 end-to-end latency?"
   * Delegates to `breakdown(measure=p95Value, groupBy=processor, segment=endToEndTurnLatency)`.
   */
  byComponent(params: LatencyByComponentParams = {}): Promise<BreakdownResponse> {
    const { percentile = "p95", ...rest } = params;
    return this.#client.breakdown({
      measure: rawValuePercentileMeasure(percentile),
      groupBy: GROUP_BY.processor,
      segment: SEGMENTS.endToEndTurnLatency,
      ...rest,
    });
  }

  /**
   * "Plot p50/p95/p99 end-to-end latency over time."
   * Deprecated compatibility path for the legacy `voice.user_to_bot_latency`
   * metric. Prefer `latency.firstResponse*` for SLA reporting.
   */
  overTime(params: LatencyOverTimeParams = {}): Promise<TimeseriesResponse> {
    const { percentiles = ["p95"], granularity, ...rest } = params;
    const percentile = percentiles[0] ?? "p95";
    return this.#client.timeseries({
      measure: rawValuePercentileMeasure(percentile),
      metricName: METRIC_NAMES.voiceUserToBotLatency,
      granularity,
      ...rest,
    });
  }

  /** Character-first, mode-specific first-response SLA summary. */
  firstResponseSummary(
    params: FirstResponseSummaryParams,
  ): Promise<FirstResponseSummaryResponse> {
    return this.#client.firstResponse.summary(params);
  }

  /** Character-first, mode-specific first-response SLA time series. */
  firstResponseOverTime(
    params: FirstResponseTimeseriesParams,
  ): Promise<FirstResponseTimeseriesResponse> {
    return this.#client.firstResponse.timeseries(params);
  }

  /** Character-first, mode-specific first-response SLA breakdown. */
  firstResponseBreakdown(
    params: FirstResponseBreakdownParams,
  ): Promise<FirstResponseBreakdownResponse> {
    return this.#client.firstResponse.breakdown(params);
  }
}
